File: packages/contracts/src/state-machines.ts
import { DocumentStatus, IncidentStatus, ReviewTaskStatus } from "./enums.js";

export interface TransitionMetadata {
  readonly actorId: string;
  readonly timestamp: Date;
  readonly reason?: string;
  readonly notes?: string;
}

export type StateTransitionsMap<T extends string> = Record<T, ReadonlySet<T>>;

export class StateTransitionError extends Error {
  constructor(public readonly errorCode: string, message: string) {
    super(message);
    this.name = "StateTransitionError";
  }
}

export class InvalidStateTransitionError extends StateTransitionError {
  constructor(
    errorCode: string,
    public readonly fromState: string,
    public readonly toState: string
  ) {
    super(errorCode, `Invalid state transition from '${fromState}' to '${toState}'.`);
    this.name = "InvalidStateTransitionError";
  }
}

export class MissingTransitionMetadataError extends StateTransitionError {
  constructor(message: string) {
    super("MISSING_TRANSITION_METADATA", message);
    this.name = "MissingTransitionMetadataError";
  }
}

export abstract class BaseStateMachine<T extends string> {
  protected abstract readonly transitions: StateTransitionsMap<T>;
  protected abstract readonly errorCode: string;

  public canTransition(fromState: T, toState: T): boolean {
    const allowed = this.transitions[fromState];
    return allowed ? allowed.has(toState) : false;
  }

  public getAllowedTransitions(fromState: T): ReadonlySet<T> {
    return this.transitions[fromState] || new Set();
  }

  public isTerminal(state: T): boolean {
    return this.getAllowedTransitions(state).size === 0;
  }

  public validateTransition(
    fromState: T,
    toState: T,
    metadata: TransitionMetadata
  ): void {
    if (!this.canTransition(fromState, toState)) {
      throw new InvalidStateTransitionError(this.errorCode, fromState, toState);
    }
    if (!metadata.actorId || metadata.actorId.trim().length === 0) {
      throw new MissingTransitionMetadataError("Actor ID is required for state transition.");
    }
    if (!metadata.timestamp) {
      throw new MissingTransitionMetadataError("Timestamp is required for state transition.");
    }
  }
}

export class DocumentStateMachineClass extends BaseStateMachine<DocumentStatus> {
  protected readonly errorCode = "DOCUMENT_INVALID_TRANSITION";
  protected readonly transitions: StateTransitionsMap<DocumentStatus> = {
    [DocumentStatus.DRAFT]: new Set([DocumentStatus.UPLOADED]),
    [DocumentStatus.UPLOADED]: new Set([DocumentStatus.PARSING]),
    [DocumentStatus.PARSING]: new Set([DocumentStatus.AI_EXTRACTED, DocumentStatus.REJECTED]),
    [DocumentStatus.AI_EXTRACTED]: new Set([DocumentStatus.IN_REVIEW]),
    [DocumentStatus.IN_REVIEW]: new Set([
      DocumentStatus.CHANGES_REQUESTED,
      DocumentStatus.REJECTED,
      DocumentStatus.SCHOLAR_REVIEW,
    ]),
    [DocumentStatus.CHANGES_REQUESTED]: new Set([DocumentStatus.IN_REVIEW]),
    [DocumentStatus.SCHOLAR_REVIEW]: new Set([
      DocumentStatus.CHANGES_REQUESTED,
      DocumentStatus.REJECTED,
      DocumentStatus.SCHOLAR_APPROVED,
    ]),
    [DocumentStatus.SCHOLAR_APPROVED]: new Set([DocumentStatus.PUBLISHED]),
    [DocumentStatus.PUBLISHED]: new Set([
      DocumentStatus.SUSPENDED,
      DocumentStatus.ARCHIVED,
      DocumentStatus.NEW_VERSION,
    ]),
    [DocumentStatus.SUSPENDED]: new Set([
      DocumentStatus.PUBLISHED,
      DocumentStatus.ARCHIVED,
      DocumentStatus.NEW_VERSION,
    ]),
    [DocumentStatus.ARCHIVED]: new Set(),
    [DocumentStatus.REJECTED]: new Set(),
    [DocumentStatus.NEW_VERSION]: new Set(),
  };

  public validateTransition(
    fromState: DocumentStatus,
    toState: DocumentStatus,
    metadata: TransitionMetadata
  ): void {
    super.validateTransition(fromState, toState, metadata);
    if (
      toState === DocumentStatus.PUBLISHED ||
      toState === DocumentStatus.SUSPENDED ||
      toState === DocumentStatus.REJECTED
    ) {
      if (!metadata.reason || metadata.reason.trim().length === 0) {
        throw new MissingTransitionMetadataError(
          `A non-empty reason is required to transition to '${toState}' status.`
        );
      }
    }
  }
}

export class ReviewTaskStateMachineClass extends BaseStateMachine<ReviewTaskStatus> {
  protected readonly errorCode = "REVIEW_TASK_INVALID_TRANSITION";
  protected readonly transitions: StateTransitionsMap<ReviewTaskStatus> = {
    [ReviewTaskStatus.OPEN]: new Set([
      ReviewTaskStatus.IN_PROGRESS,
      ReviewTaskStatus.COMPLETED,
      ReviewTaskStatus.CANCELLED,
    ]),
    [ReviewTaskStatus.IN_PROGRESS]: new Set([
      ReviewTaskStatus.COMPLETED,
      ReviewTaskStatus.CANCELLED,
    ]),
    [ReviewTaskStatus.COMPLETED]: new Set(),
    [ReviewTaskStatus.CANCELLED]: new Set(),
  };
}

export class IncidentStateMachineClass extends BaseStateMachine<IncidentStatus> {
  protected readonly errorCode = "INCIDENT_INVALID_TRANSITION";
  protected readonly transitions: StateTransitionsMap<IncidentStatus> = {
    [IncidentStatus.OPEN]: new Set([IncidentStatus.TRIAGED, IncidentStatus.CLOSED]),
    [IncidentStatus.TRIAGED]: new Set([IncidentStatus.MITIGATED, IncidentStatus.CLOSED]),
    [IncidentStatus.MITIGATED]: new Set([IncidentStatus.RESOLVED, IncidentStatus.CLOSED]),
    [IncidentStatus.RESOLVED]: new Set([IncidentStatus.CLOSED]),
    [IncidentStatus.CLOSED]: new Set([IncidentStatus.OPEN]),
  };
}

export const documentStateMachine = new DocumentStateMachineClass();
export const reviewTaskStateMachine = new ReviewTaskStateMachineClass();
export const incidentStateMachine = new IncidentStateMachineClass();

File: services/common/src/zayd_common/state_machines.py
from datetime import UTC, datetime
from typing import Any, ClassVar, cast

from pydantic import BaseModel, Field

from zayd_common.enums import DocumentStatus, IncidentStatus, ReviewTaskStatus
from zayd_common.exceptions import InvalidStateTransitionError, MissingTransitionMetadataError


class TransitionMetadata(BaseModel):
    """Metadata recorded with every state transition for auditing and compliance."""

    actor_id: str = Field(
        ...,
        min_length=1,
        description="Unique identifier of the actor performing the transition",
    )
    timestamp: datetime = Field(
        default_factory=lambda: datetime.now(UTC),
        description="Timestamp of when the transition occurred",
    )
    reason: str | None = Field(
        default=None,
        description="Detailed reason for the transition, required for sensitive states",
    )
    notes: str | None = Field(
        default=None,
        description="Optional additional notes or context",
    )


class BaseStateMachine[T]:
    """Generic base class for status lifecycle and state transition rules."""

    TRANSITIONS: ClassVar[dict[Any, set[Any]]] = {}
    ERROR_CODE: ClassVar[str] = "INVALID_TRANSITION"

    @classmethod
    def can_transition(cls, from_state: T, to_state: T) -> bool:
        """Return whether a transition from from_state to to_state is allowed."""
        return to_state in cls.TRANSITIONS.get(from_state, set())

    @classmethod
    def get_allowed_transitions(cls, from_state: T) -> set[T]:
        """Return the set of all allowed states transitioning from from_state."""
        return cast(set[T], cls.TRANSITIONS.get(from_state, set()))

    @classmethod
    def is_terminal(cls, state: T) -> bool:
        """Return True if the state has no outgoing transitions."""
        return not cls.get_allowed_transitions(state)

    @classmethod
    def validate_transition(cls, from_state: T, to_state: T, metadata: TransitionMetadata) -> None:
        """Validate if transition is allowed and metadata is complete.

        Raises InvalidStateTransitionError or MissingTransitionMetadataError.
        """
        if not cls.can_transition(from_state, to_state):
            raise InvalidStateTransitionError(cls.ERROR_CODE, str(from_state), str(to_state))

        if not metadata.actor_id or not metadata.actor_id.strip():
            raise MissingTransitionMetadataError("Actor ID is required for state transition.")

        if not metadata.timestamp:
            raise MissingTransitionMetadataError("Timestamp is required for state transition.")


class DocumentStateMachine(BaseStateMachine[DocumentStatus]):
    """State machine governing Document status lifecycles."""

    ERROR_CODE: ClassVar[str] = "DOCUMENT_INVALID_TRANSITION"
    TRANSITIONS: ClassVar[dict[DocumentStatus, set[DocumentStatus]]] = {
        DocumentStatus.DRAFT: {DocumentStatus.UPLOADED},
        DocumentStatus.UPLOADED: {DocumentStatus.PARSING},
        DocumentStatus.PARSING: {DocumentStatus.AI_EXTRACTED, DocumentStatus.REJECTED},
        DocumentStatus.AI_EXTRACTED: {DocumentStatus.IN_REVIEW},
        DocumentStatus.IN_REVIEW: {
            DocumentStatus.CHANGES_REQUESTED,
            DocumentStatus.REJECTED,
            DocumentStatus.SCHOLAR_REVIEW,
        },
        DocumentStatus.CHANGES_REQUESTED: {DocumentStatus.IN_REVIEW},
        DocumentStatus.SCHOLAR_REVIEW: {
            DocumentStatus.CHANGES_REQUESTED,
            DocumentStatus.REJECTED,
            DocumentStatus.SCHOLAR_APPROVED,
        },
        DocumentStatus.SCHOLAR_APPROVED: {DocumentStatus.PUBLISHED},
        DocumentStatus.PUBLISHED: {
            DocumentStatus.SUSPENDED,
            DocumentStatus.ARCHIVED,
            DocumentStatus.NEW_VERSION,
        },
        DocumentStatus.SUSPENDED: {
            DocumentStatus.PUBLISHED,
            DocumentStatus.ARCHIVED,
            DocumentStatus.NEW_VERSION,
        },
        DocumentStatus.ARCHIVED: set(),
        DocumentStatus.REJECTED: set(),
        DocumentStatus.NEW_VERSION: set(),
    }

    @classmethod
    def validate_transition(
        cls,
        from_state: DocumentStatus,
        to_state: DocumentStatus,
        metadata: TransitionMetadata,
    ) -> None:
        super().validate_transition(from_state, to_state, metadata)

        # Transition to published, suspended, or rejected requires non-empty reason
        if to_state in (
            DocumentStatus.PUBLISHED,
            DocumentStatus.SUSPENDED,
            DocumentStatus.REJECTED,
        ):
            if not metadata.reason or not metadata.reason.strip():
                raise MissingTransitionMetadataError(
                    f"A non-empty reason is required to transition to '{to_state}' status."
                )


class ReviewTaskStateMachine(BaseStateMachine[ReviewTaskStatus]):
    """State machine governing Review Task status lifecycles."""

    ERROR_CODE: ClassVar[str] = "REVIEW_TASK_INVALID_TRANSITION"
    TRANSITIONS: ClassVar[dict[ReviewTaskStatus, set[ReviewTaskStatus]]] = {
        ReviewTaskStatus.OPEN: {
            ReviewTaskStatus.IN_PROGRESS,
            ReviewTaskStatus.COMPLETED,
            ReviewTaskStatus.CANCELLED,
        },
        ReviewTaskStatus.IN_PROGRESS: {
            ReviewTaskStatus.COMPLETED,
            ReviewTaskStatus.CANCELLED,
        },
        ReviewTaskStatus.COMPLETED: set(),
        ReviewTaskStatus.CANCELLED: set(),
    }


class IncidentStateMachine(BaseStateMachine[IncidentStatus]):
    """State machine governing Incident status lifecycles."""

    ERROR_CODE: ClassVar[str] = "INCIDENT_INVALID_TRANSITION"
    TRANSITIONS: ClassVar[dict[IncidentStatus, set[IncidentStatus]]] = {
        IncidentStatus.OPEN: {IncidentStatus.TRIAGED, IncidentStatus.CLOSED},
        IncidentStatus.TRIAGED: {IncidentStatus.MITIGATED, IncidentStatus.CLOSED},
        IncidentStatus.MITIGATED: {IncidentStatus.RESOLVED, IncidentStatus.CLOSED},
        IncidentStatus.RESOLVED: {IncidentStatus.CLOSED},
        IncidentStatus.CLOSED: {IncidentStatus.OPEN},
    }
